injector/injector.go

```diff
@@ -0,0 +1,340 @@
+	if vastXML == "" && NURL == "" {
+		// TODO Log a adapter.<bidder-name>.requests.badserverresponse
+		return vastXML
+	}
+
+	if vastXML == "" {
+		return fmt.Sprintf(emptyAdmResponse, NURL)
+	}
+
+	var outputXML strings.Builder
+	encoder := xml.NewEncoder(&outputXML)
+	state := &InjectionState{
+		injectTracker:         false,
+		injectVideoClicks:     false,
+		inlineWrapperTagFound: false,
+		wrapperTagFound:       false,
+		impressionTagFound:    false,
+		errorTagFound:         false,
+		creativeId:            "",
+		isCreative:            false,
+		companionTagFound:     false,
+		nonLinearTagFound:     false,
+	}
+
+	reader := strings.NewReader(vastXML)
+	decoder := xml.NewDecoder(reader)
+
+	for {
+		rawToken, err := decoder.RawToken()
+		if err != nil {
+			if err == io.EOF {
+				break
+			}
+			return ""
+		}
+
+		switch token := rawToken.(type) {
+		case xml.StartElement:
+			trackerinjector.handleStartElement(token, state, &outputXML, encoder)
+		case xml.EndElement:
+			trackerinjector.handleEndElement(token, state, &outputXML, encoder)
+		case xml.CharData:
+			charData := strings.Trim(string(token), trimRunes)
+			if len(charData) != 0 {
+				encoder.Flush()
+				outputXML.WriteString("<![CDATA[")
+				outputXML.WriteString(charData)
+				outputXML.WriteString("]]>")
+				continue
+			}
+		default:
+			encoder.EncodeToken(rawToken)
+		}
+	}
+
+	encoder.Flush()
+
+	if !state.inlineWrapperTagFound {
+		// Todo log adapter.<bidder-name>.requests.badserverresponse metrics
+		return vastXML
+	}
+	return outputXML.String()
+}
+
+func (trackerinjector *TrackerInjector) handleStartElement(token xml.StartElement, state *InjectionState, outputXML *strings.Builder, encoder *xml.Encoder) {
+	switch token.Name.Local {
+	case wrapperCase:
+		state.wrapperTagFound = true
+		encoder.EncodeToken(token)
+	case creativeCase:
+		state.isCreative = true
+		for _, attr := range token.Attr {
+			if strings.ToLower(attr.Name.Local) == adId {
+				state.creativeId = attr.Value
+			}
+		}
+		encoder.EncodeToken(token)
+	case linearCase:
+		state.injectVideoClicks = true
+		state.injectTracker = true
+		encoder.EncodeToken(token)
+	case videoClicksCase:
+		state.injectVideoClicks = false
+		encoder.EncodeToken(token)
+		encoder.Flush()
+		trackerinjector.addClickTrackingEvent(outputXML, state.creativeId, false)
+	case nonLinearAdsCase:
+		state.injectTracker = true
+		encoder.EncodeToken(token)
+	case trackingEventsCase:
+		if state.isCreative {
+			state.injectTracker = false
+			encoder.EncodeToken(token)
+			encoder.Flush()
+			trackerinjector.addTrackingEvent(outputXML, state.creativeId, false)
+		}
+	default:
+		encoder.EncodeToken(token)
+	}
+}
+
+func (trackerinjector *TrackerInjector) handleEndElement(token xml.EndElement, state *InjectionState, outputXML *strings.Builder, encoder *xml.Encoder) {
+	switch token.Name.Local {
+	case impressionCase:
+		fmt.Println(outputXML.String())
```

Comment:
Nitpick: Can you remove this print statement.

Reply:
Done